ansible_builder/containerfile.py

```diff
@@ -63,27 +63,93 @@ def prepare(self):
+                self.steps.append(f'ARG PYPKG={python}')
+                # FIXME: better dnf cleanup needed?
+                self.steps.append('RUN dnf install $PYPKG -y && dnf clean all')
+
+            if self.definition.ansible_ref_install_list:
+                self.steps.append('ARG ANSIBLE_INSTALL_REFS')
+                self.steps.append('ARG PYCMD')
+                self.steps.append('RUN $PYCMD -m ensurepip && $PYCMD -m pip install --no-cache-dir $ANSIBLE_INSTALL_REFS')
+
+        self._create_folder_copy_files()
+        self._insert_custom_steps('append_base')
+
+        ######################################################################
+        # First stage (aka, galaxy): install roles/collections
+        ######################################################################
+
+        self.steps.extend([
+            "",
+            "FROM base as galaxy",
+            "ARG ANSIBLE_GALAXY_CLI_COLLECTION_OPTS={}".format(
+                self.definition.build_arg_defaults['ANSIBLE_GALAXY_CLI_COLLECTION_OPTS']
+            ),
+            "ARG ANSIBLE_GALAXY_CLI_ROLE_OPTS={}".format(
+                self.definition.build_arg_defaults['ANSIBLE_GALAXY_CLI_ROLE_OPTS']
+            ),
+            "USER root",
+            ""
+        ])
+
+        self._insert_custom_steps('prepend_galaxy')
+        self._prepare_ansible_config_file()
+        self._prepare_build_context()
+        self._prepare_galaxy_install_steps()
+        self._insert_custom_steps('append_galaxy')
+
+        ######################################################################
+        # Second stage (aka, builder): assemble (pip installs, bindep run)
+        ######################################################################
+
+        if self.definition.builder_image:
+            self.steps.extend([
+                "",
+                "FROM $EE_BUILDER_IMAGE as builder"
+                "",
+            ])
+        else:
+            # dynamic builder, create from customized base
+            self.steps.extend([
+                'FROM base as builder',
+                'ARG PYCMD',
+                'RUN $PYCMD -m pip install --no-cache-dir bindep pyyaml requirements-parser'
+            ])
+
+        self._insert_custom_steps('prepend_builder')
+        self._prepare_galaxy_copy_steps()
+        self._prepare_introspect_assemble_steps()
+        self._insert_custom_steps('append_builder')
+
+        ######################################################################
+        # Final stage: package manager installs from bindep output
+        ######################################################################
+
+        self.steps.extend([
+            "",
+            "FROM base",
+            "USER root",
+            "ARG PYCMD"  # this is consumed as an envvar by the install-from-bindep script
+            "",
+        ])
+
+        self._insert_custom_steps('prepend')  # EE v1
```

Comment:
In cases where there's a direct replacement in a newer schema, does it make sense to just implement on the new version? (eg, maybe a "schema upgrade" step after validation that makes it so we don't need to care about old/new differences in things like this except where there's an actual behavioral difference)

I can live with either way, just something to think about.

Reply:
I'll need to think about this one a bit. Currently, depending on which schema version is used, one of these is turned into a no-op because the validation up front guarantees that only one will be present.